lib/app/utils.dart

```diff
@@ -61,44 +68,64 @@ Future<bool> tryLaunchingUrl(String url) async {
   return false;
 }
 
+T tryToParse<T>(T Function() parser, {@required T defaultValue}) {
+  try {
+    return parser();
+  } catch (_) {
+    return defaultValue;
+  }
+}
+
 /// An error indicating that a permission wasn't granted by the user.
 class PermissionNotGranted<T> implements Exception {
   String toString() => "A permission wasn't granted by the user.";
 }
 
+class Id<T> {
+  final String id;
+
+  Id(this.id);
+
+  String toString() => id;
+}
+
 /// A special kind of item that also carries its id.
 abstract class Entity {
-  Id<Entity> get id;
+  Id get id;
   const Entity();
 }
 
-abstract class CollectionDownloader<Item extends Entity>
-    extends Repository<Item> {
-  Future<List<Item>> _downloader;
-  Map<Id<Item>, Item> _items;
-
-  CollectionDownloader() : super(isFinite: true, isMutable: false);
-
-  Future<void> _ensureItemsAreDownloaded() async {
-    _downloader ??= downloadAll();
-    _items ??= {for (var item in await _downloader) item.id: item};
-  }
-
-  @override
-  Stream<Item> fetch(Id<Item> id) async* {
-    await _ensureItemsAreDownloaded();
-    if (_items.containsKey(id)) {
-      yield _items[id];
-    } else {
-      throw ItemNotFound(id);
-    }
-  }
-
-  @override
-  Stream<Map<Id<Item>, Item>> fetchAll() async* {
-    await _ensureItemsAreDownloaded();
-    yield _items;
-  }
+class HiveCacheController<Item extends Entity>
+    extends CacheController<List<Item>> {
+  final StorageService storage;
+  final String parentKey;
+
+  HiveCacheController({
+    @required this.storage,
+    @required this.parentKey,
+    Future<List<Item>> Function() fetcher,
+  })  : assert(storage != null),
+        assert(parentKey != null),
+        super(
+          saveToCache: (items) async {
+            for (var item in items) {
+              await storage.cache.put(item.id.id, parentKey, item);
+            }
+          },
+          loadFromCache: () async {
+            return await storage.cache.getChildrenOfType<Item>(parentKey)
+              ..sort();
+          },
+          fetcher: () async {
+            return await fetcher()
+              ..sort();
+          },
+        );
+}
 
-  Future<List<Item>> downloadAll();
+Future<void> logOut(BuildContext context) async {
+  await Provider.of<StorageService>(context).clear();
+  Navigator.of(context, rootNavigator: true).pushReplacement(TopLevelPageRoute(
+    builder: (_) => LoginScreen(),
+  ));
```

Comment:
Hmm, I agree that the logic is login-related, so the `await Provider.of<StorageService>(context).clear();` could indeed be put into the `MeService`. 😊
But the `Navigator....` part is UI-logic and I'm not sure we should mix that into the domain-specific logic.
And because it's used in various places (like the `ErrorScreen` for a `NotAuthenticatedError` or the `Menu` for the logout button.